staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,1236 @@
+            trace!("Error::UnexpectedDescriptorCount");
+            return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+        }
+
+        let desc_response = descriptors[0];
+        if !desc_response.is_write_only() {
+            return Err(Error::UnexpectedReadableDescriptor(1));
+        }
+
+        let mut response = match self.controller.write().unwrap().pop() {
+            Ok(item) => item,
+            Err(_) => return Err(Error::HandleEventUnknown),
+        };
+
+        CanController::print_can_frame(response);
+
+        if (response.can_id.to_native() & CAN_ERR_FLAG) != 0 {
+            if (response.can_id.to_native() & CAN_ERR_BUSOFF) != 0 {
+                //TODO: vcan->busoff = true;
+                warn!("Got BusOff error frame, device does a local bus off\n");
+            } else {
+                trace!("Dropping error frame 0x{:x}\n", response.can_id.to_native());
+            }
+            return Ok(true);
+        }
+
+        let mut can_rx = VirtioCanFrame::default();
+        can_rx.msg_type = VIRTIO_CAN_RX.into();
+        can_rx.can_id = response.can_id;
+        can_rx.length = response.length;
+        can_rx.flags = (can_rx.flags.to_native() | VIRTIO_CAN_FLAGS_FD).into();
+
+        if (response.flags.to_native() & CAN_FRMF_TYPE_FD) != 0 {
+            if !self.check_features(VIRTIO_CAN_F_CAN_FD) {
+                warn!("Drop non-supported CAN FD frame");
+                return Err(Error::UnexpectedFdFlag);
+            }
+        } else if !self.check_features(VIRTIO_CAN_F_CAN_CLASSIC) {
+            warn!("Drop non-supported CAN classic frame");
+            return Err(Error::UnexpectedClassicFlag);
+        }
+        if (response.can_id.to_native() & CAN_RTR_FLAG) != 0
+            && !self.check_features(VIRTIO_CAN_F_RTR_FRAMES)
+        {
+            warn!("Drop non-supported RTR frame");
+            return Err(Error::UnexpectedRtrFlag);
+        }
+
+        if (response.can_id.to_native() & CAN_EFF_FLAG) != 0 {
+            can_rx.flags = VIRTIO_CAN_FLAGS_EXTENDED.into();
+            can_rx.can_id = (response.can_id.to_native() & CAN_EFF_MASK).into();
+        } else {
+            can_rx.can_id = (response.can_id.to_native() & CAN_SFF_MASK).into();
+        }
+        if (response.can_id.to_native() & CAN_RTR_FLAG) != 0 {
+            can_rx.flags = (can_rx.flags.to_native() & VIRTIO_CAN_FLAGS_RTR).into();
+        }
+
+        // Treat Vcan interface as CANFD if MTU is set to 64 bytes.
+        //
+        // Vcan can not be configured as CANFD interface, but it is
+        // possible to configure its MTU to 64 bytes. So if a messages
+        // bigger than 8 bytes is being received we consider it as
+        // CANFD message.
+        let can_in_name = self.controller.read().unwrap().can_in_name.clone();
+        if self.check_features(VIRTIO_CAN_F_CAN_FD)
+            && response.length.to_native() > 8
+            && can_in_name == "vcan0"
```

Comment:
Yes of course, that's a hacky point.

By defalut, `vcan` supports CAN frame and not CANFD, so we can use `vcan0` to send CAN frames with 8 bytes in the data field. If we try to set it up as a CANFD device by running `sudo ip link set vcan0 type can bitrate 1000000 dbitrate 5000000 fd on`, this will fails.

But there is a workaround to make `vcan` accept and send CAN frames with the data field having 0 to 64 bytes by setting the MTU to 72 bytes (`sudo ip link set vcan0 mtu 72`). But even in that case, `vcan0` does not support fully FD frames, since the received frame would not include `CAN_FRMF_TYPE_FD` under flags section.

So in our scenario, this means that, on host side, if we try to run `cangen -f vcan0`, the guest side (`candump can0`) will read only the 8 first bytes of the data (because `CAN_FRMF_TYPE_FD` is missing from the frames' flags). Instead, by adding the above code, and eventually `CAN_FRMF_TYPE_FD` flag on each received frame, when using "vcan0" interface on the host, the guest will read (`candump can0`) CANFD frames with data till 64 bytes.

To complete this answer, it is worth to mention that this practice takes place only while using "vcan0" device, in order to make it work as CANFD device. This was added most of all to give the ability to test both CAN and CANFD cases without the need of any HW CANFD device connected on host.

As for HW CANFD devices (connected on the host), frames received by vhost-device-can (from the host's `canX` interface, with the intention to be written into guests virtqueues) SHOULD already include the above flag, otherwise the guest SHOULD and WILL treat those frame as simple CAN ones.